### PR TITLE
YAML writing in python

### PR DESCRIPTION
**Pull request by tilmantroester**
This PR removes the C-level yaml reader and writer routines. The reader wasn't being used and the writer has now been replaced by a python level reader. This makes it easier to add new parameters in the future. 
The C-level writer wasn't consistent anymore in any case (e.g., C-level wrote `z_mgrowth`, while Python-level read `z_mg`).

The PR also allows file pointers and streams to be used instead of file names. This enables passing around information without need to write to disk.

Removed `df_mg` and `z_mg` from the yaml reader, in anticipation of their eventual removal (#842).

### CODE REVIEW

File: pyccl/core.py [hunk 1]
```diff
@@ -229,27 +229,53 @@ def write_yaml(self, filename):
...
-            params = yaml.load(fp, Loader=yaml.Loader)
+        if isinstance(filename, str):
+            with open(filename, 'r') as fp:
+                params = yaml.load(fp, Loader=yaml.Loader)
+        else:
+            params = yaml.load(filename, Loader=yaml.Loader)
+
+        # Check for missing values in yaml files created with the C-level
+        # yaml dumper
+        if any([p not in params for p in ["m_nu", "m_nu_type"]]):
```
Comment (damonge) [inline pyccl/core.py:272]: Is this the only case that needs to be checked?

**Review by damonge — COMMENTED** [review]
So nice to see unneeded code go. thanks! Just one comment, and I'll approve as soon as the tests pass

File: pyccl/core.py (hunk 1 above)
Comment (tilmantroester) [reply pyccl/core.py:272]: Technically `z_mg` and `df_mg` as well but I don't think that had ever worked (C-level calls them `z_mgrowth`). The new MG parameters (`c1_mg`, etc) don't have a fall-back either, so the readers already fail on yaml files created with old versions of CCL. 
So I'm tempted to replace that check with just a check for a `KeyError` and give a warning if any parameters are missing.

Comment (tilmantroester) [reply pyccl/core.py]: The current version should handle this more cleanly and robustly.

**Comment by tilmantroester** [comment]
This should now handle missing values much better and address #817.

**Review by damonge — APPROVED** [review]
LGTM, but would be good to check if this messes up any of what @joezuntz has been doing

**Comment by joezuntz** [comment]
This is fine from my end - we only use the python interface.  I might need to regenerate one file but that's all.

**Comment by damonge** [comment]
Closes #817

**Pull request merged**